framework/aster-frame/src/vm/page_table/mod.rs

```diff
@@ -0,0 +1,393 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{boxed::Box, sync::Arc};
+use core::{fmt::Debug, marker::PhantomData, ops::Range};
+
+use crate::{
+    arch::mm::{activate_page_table, PageTableConsts, PageTableEntry},
+    sync::SpinLock,
+    vm::{paddr_to_vaddr, Paddr, Vaddr, VmAllocOptions, VmFrame, VmFrameVec, PAGE_SIZE},
+};
+
+mod properties;
+pub use properties::*;
+mod cursor;
+use cursor::*;
+#[cfg(ktest)]
+mod test;
+
+#[derive(Debug)]
+pub enum PageTableError {
+    InvalidVaddr(Vaddr),
+    InvalidVaddrRange(Range<Vaddr>),
+    VaddrNotAligned(Vaddr),
+    VaddrRangeNotAligned(Range<Vaddr>),
+    PaddrNotAligned(Paddr),
+    PaddrRangeNotAligned(Range<Paddr>),
+    // Protecting a mapping that does not exist.
+    ProtectInvalid,
+}
+
+/// This is a compile-time technique to force the frame developers to distinguish
+/// between the kernel global page table instance, process specific user page table
+/// instance, and device page table instances.
+pub trait PageTableMode {
+    /// The range of virtual addresses that the page table can manage.
+    const VADDR_RANGE: Range<Vaddr>;
+}
+
+#[derive(Clone)]
+pub struct UserMode {}
+
+impl PageTableMode for UserMode {
+    const VADDR_RANGE: Range<Vaddr> = 0..super::MAX_USERSPACE_VADDR;
+}
+
+#[derive(Clone)]
+pub struct KernelMode {}
+
+impl PageTableMode for KernelMode {
+    const VADDR_RANGE: Range<Vaddr> = super::KERNEL_BASE_VADDR..super::KERNEL_END_VADDR;
+}
+
+/// A page table instance.
+pub struct PageTable<
+    M: PageTableMode,
+    E: PageTableEntryTrait = PageTableEntry,
+    C: PageTableConstsTrait = PageTableConsts,
+> where
+    [(); nr_entries_per_frame::<C>()]:,
+    [(); C::NR_LEVELS]:,
+{
+    root_frame: PtfRef<E, C>,
+    _phantom: PhantomData<M>,
+}
+
+/// A page table frame.
+/// It's also frequently referred to as a page table in many architectural documentations.
+#[derive(Debug)]
+struct PageTableFrame<E: PageTableEntryTrait, C: PageTableConstsTrait>
+where
+    [(); nr_entries_per_frame::<C>()]:,
+    [(); C::NR_LEVELS]:,
+{
+    pub inner: VmFrame,
+    #[allow(clippy::type_complexity)]
+    pub child: Option<Box<[Option<PtfRef<E, C>>; nr_entries_per_frame::<C>()]>>,
```

Comment:
This is essentially a copy of page tables, but represented in Rust structures, right? I do not see too much value in why we need to maintain this, especially after we manage to avoid the abuse of internal `Arc<_>`s and `SpinLock<_>`s.

Reply:
Yes it is a copy of page tables. I don't really want to introduce it in the first place. However, we need a data structure to keep track of the page table frames, which would be allocated and freed dynamically. A PTE can keep the pointer to the frame but it cannot own the frame due to technical restrictions.

The rationale for `Arc` is that the kernel page tables are shared.

Reply:
> The rationale for Arc is that the kernel page tables are shared.

Well, that's indeed a very strong argument for the `Arc<SpinLock<_>>` structure. I simply forgot about it thing when I had a quick look at this PR. Thanks for pointing that out.

But if this is the case, I think that you need to take all the parent locks when you operate on the leaf. Otherwise, race conditions will lead to inconsistent results, won't they?